### PR TITLE
Fiber-safe ConnectionPool

### PR DESCRIPTION
**Pull request by machty**
### Summary

A stab at a Fiber-safe AR ConnectionPool. I had to make a few tweaks to IsolationExecutionState so that I could implement the various cache key methods in the pool and decouple them from Thread.current.

I haven't added tests yet, but I've played with this locally and it seems to be working.

### Other Information

Please see here for a discussion for why a Fiber-safe ConnectionPool is desirable: https://github.com/rails/rails/issues/42271

Note that this approach would also make ActiveRecord within Sidekiq fiber-safe, along with any other gem that uses `app.reloader.wrap` to wrap requests/jobs/etc.

### TODO

- [x] Discuss this approach / get some feedback
- [x] Tests

### CODE REVIEW

File: activesupport/lib/active_support/isolated_execution_state.rb [hunk 1]
```diff
@@ -19,8 +19,8 @@ def isolation_level=(level)
 
         if level != isolation_level
           clear
-          singleton_class.alias_method(:current, "current_#{level}")
-          singleton_class.send(:private, :current)
+          singleton_class.alias_method(:current_object, "current_#{level}")
```
Comment (byroot) [inline activesupport/lib/active_support/isolated_execution_state.rb:22]: Why `current_object` instead of keeping `current`?

File: activesupport/lib/active_support/isolated_execution_state.rb [hunk 2]
```diff
@@ -19,8 +19,8 @@ def isolation_level=(level)
...
+          # singleton_class.send(:private, :current)
```
Comment (byroot) [inline activesupport/lib/active_support/isolated_execution_state.rb:23]:
```suggestion
```

**Review by byroot — COMMENTED** [review]
Looks mostly good so far. Let's see with tests.

Also I believe that a common demand was for checking the connection back in whenever it's not in use so that you don't need as many connections as you have thread / fibers, but It can perfectly be a followup.

File: activesupport/lib/active_support/isolated_execution_state.rb (hunk 1 above)
Comment (machty) [reply activesupport/lib/active_support/isolated_execution_state.rb:22]: `current` points to the Hash that gets [`||=` initialized](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/isolated_execution_state.rb#L46) either on the Fiber or Thread, depending on the selected isolation menu. It wasn't clear how/whether I could use this Hash object as the cache key within ConnectionPool logic; the hash itself is mutable which disqualifies it for use as a cache key, and it would be weird to use it that way anyway; could I perhaps use the `.object_id` of that hash, i.e. `current.object_id`?

Comment (byroot) [reply activesupport/lib/active_support/isolated_execution_state.rb:22]: > current points to the Hash ...

Sorry, my comment wasn't clear. I was suggesting to have `current` as the public method, the other one is private so the name doesn't matter that much. 

But now that I took a second look at the code:

> It wasn't clear how/whether I could use this Hash object as the cache key ...

Wouldn't `IsolatedExecutionState.unique_id` do the trick? The advantage being that you wouldn't hold a reference on thread / fiber, so that would avoid a bunch of leak scenarios. The only downside would be the few `owned by a different thread:` errors, but we could have the `Thread / Fiber #inspect` in that `unique_id`.

Comment (byroot) [reply activesupport/lib/active_support/isolated_execution_state.rb:22]: > but we could have the Thread / Fiber #inspect in that unique_id.

Or have a WeakRef to the thread / fiber.

Comment (machty) [reply activesupport/lib/active_support/isolated_execution_state.rb:22]: I forgot to mention: even if we can use `current.object_id`, the reaper checks the `alive?` method of the connection owner in order to detect whether a Thread (or Fiber) has finished executing. I think this pushes in the direction of exposing `current_object`. i guess we could also consider introducing some intermediary class that the Fiber or Thread could generate that is consulted throughout the process of checking out a connection and checking if the owner is still alive, but it feels a bit overcomplicated.

Comment (byroot) [reply activesupport/lib/active_support/isolated_execution_state.rb:22]: > the reaper checks the `alive?` method of the connection owner in order to detect whether a Thread (or Fiber) has finished executing. I think this pushes in the direction of exposing current_object

Hum, I missed that. Ok, so indeed we'll have to expose it. I'm really not fan of `current_object` though, could be `current`, `current_owner`, `current_executor` or something like that, but `current_object` is really not telling much.

**Comment by machty** [comment]
@byroot I changed a few things around, ultimately renaming `current_object` to `executor`, but I'm not loving that because `executor` makes me think of ThreadPoolExecutor, i.e. an "executor" is already defined as a different kind of abstraction that might internally use threads or fibers. But it's easy enough to substitute the name at a later point; in the meantime I wanted to raise another concern:

IsolatedExecutionState seemed to be a step in the direction of abstracting whether the "current running thing" is a Thread or a Fiber, but as I look around the Rails codebase, I see lots of code that is hard-wired to a Thread-centric implementation. The more we push in the direction of making Rails Async/Fiber-friendly, the more we'll run into framework code that's hardwired to Thread-centric APIs, some examples:

- [ActionCable::Connection::StreamEventLoop uses a Thread and ThreadPoolExecutor to implement an event loop](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/actioncable/lib/action_cable/connection/stream_event_loop.rb#L69-L75)
- [The ActionCable PostgreSQL subscription adapter listens to async events via a thread running in a loop](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/actioncable/lib/action_cable/subscription_adapter/postgresql.rb#L82-L85)
- [The ActionCable Redis subscription handler does a similar thing](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/actioncable/lib/action_cable/subscription_adapter/redis.rb#L150-L155)
- [ActionController::Live spins up a thread to stream the response](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/actionpack/lib/action_controller/metal/live.rb#L339-L345)
- [ActiveJob's in-process AsyncAdapter uses a ThreadPoolExecutor](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/activejob/lib/active_job/queue_adapters/async_adapter.rb#L89)
  - this maybe fine; we could probably just create another Fiber-friendly adapter?
- [load_async is driven either by a global or per connect ThreadPoolExecutor](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/activerecord/lib/active_record.rb#L219)
- [ActiveRecord::ConnectionAdapters::TransactionState check Thread#status == 'aborting' and there is no Fiber#status](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L336)

...and more.

I'm wondering if maybe `IsolatedExecutionState` needs to evolve into something like `ConcurrencyAdapter`, a global configurable object can be overridden to support Fiber-y libraries like async. Such as class would have hooks for scheduling async operations, and perhaps allocating/building concurrency primitives like Fiber/Thread pool executors.

Maybe it's premature to think about this, but I worry that there might be more features in the pipeline that are Thread-centric, and without something like a `ConcurrencyAdapter` in place, we're just making it harder on ourselves to Fiber-ize/Async-ify Rails in the future?

**Comment by byroot** [comment]
> Maybe it's premature to think about this

I think it is yes. Action Cable etc are decoupled enough from the core of Rails that's it's an entirely different matter to me.

As for `load_async`, yes it uses a thread pool, but that's due to the limitation of most db clients, I have some thoughts on how to do this without a thread pool, but it's quite a lot of work and I expect to run into lots of issues.

Also overall I don't think being friendly with Fiber based servers mean we need to eliminate all threads, the two can co-exist as far as I know, and things like say the connection reaper still make sense as an independently scheduled thread.

So yeah, we can build that kind of abstraction if they are needed, but let's take things in order.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 3]
```diff
@@ -223,17 +223,18 @@ def valid_type?(type) # :nodoc:
       # this method must only be called while holding connection pool's mutex
       def lease
         if in_use?
-          msg = +"Cannot lease connection, "
-          if @owner == Thread.current
+          msg = + "Cannot lease connection, "
+          # if @owner == Thread.current
```
Comment (byroot) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:227]:
```suggestion
```

No commented out code please.

**Comment by byroot** [comment]
Please squash you commits together, as besides a few exceptions, in general PRs should be a single commit.

Also do you think there's a way you could meaningfully test this? E.g. have a few connection pool tests that switch the execution level and make sure the connection pool behaves as expected?

**Comment by ioquatix** [comment]
The fiber scheduler is capable of event driven scheduling of thread primitives and operations so you can mix the two together, or incrementally make improvements.

`pg` now supports event driven concurrency to a certain extent so that will solve some of the DB related pains.

Regarding `IsolatedExecutionContext` in Ruby 3.2 I want to try to introduce a general ExecutionContext model with inheritable state similar to Kotlin's coroutine context. I believe it's the next step to help people deal with `Thread.current` sprinkled throughout the code, by providing a better model.

**Comment by machty** [comment]
> Regarding `IsolatedExecutionState`, in Ruby 3.2 I want to try to introduce a general `ExecutionContext` model with inheritable state similar to Kotlin's coroutine context. I believe it's the next step to help people deal with `Thread.current` sprinkled throughout the code, by providing a better model.

Would this also address the corner cases involving Enumerator and other internal Ruby fibers?

**Comment by machty** [comment]
> As for `load_async`, yes it uses a thread pool, but that's due to the limitation of most db clients, I have some thoughts on how to do this without a thread pool, but it's quite a lot of work and I expect to run into lots of issues.

@byroot are you perhaps referring to something like pipeline mode in Postgres 14?

**Comment by byroot** [comment]
> perhaps referring to something like pipeline mode in Postgres 14?

No, I'm referring to the db client async APIs:

  - https://dev.mysql.com/doc/c-api/8.0/en/c-api-asynchronous-interface.html
  - https://www.postgresql.org/docs/8.3/libpq-async.html

Which would allow to send the query synchronously first, and then have a background thread doing basically `IO.select` to get the results ASAP and put the connection back in the pool. So instead of a thread pool, we'd have a single background thread.

The C clients expose these async API, but the support in the Ruby client is not quite ideal (based on my initial investigation, might work better than I expect).

**Comment by ioquatix** [comment]
> Would this also address the corner cases involving Enumerator and other internal Ruby fibers?

Yes.

> No, I'm referring to the db client async APIs.

The https://github.com/socketry/db already implements these interfaces (well, it prefers `libmariadbclient` which has a better async interface for `mysql`). It works pretty well.

File: activerecord/test/cases/connection_pool_test.rb [hunk 4]
```diff
@@ -5,11 +5,15 @@
 
 module ActiveRecord
   module ConnectionAdapters
-    class ConnectionPoolTest < ActiveRecord::TestCase
+    module ConnectionPoolTests
```
Comment (machty) [inline activerecord/test/cases/connection_pool_test.rb:8]: I reworked this file so that there are two ActiveRecord::TestCase subclasses that include these shared tests as a module with slightly different test setup and some overrides as to how threads (or fibers) are spun up, joined, `pass`'d to. This approach achieves decent test thoroughness while keeping the diff small, and it highlights a few things we might want to come back to, such as testing this more thoroughly in the Future with a Ruby 3.1+ Fiber Scheduler.

File: activerecord/test/cases/connection_pool_test.rb [hunk 5]
```diff
@@ -491,19 +500,16 @@ def test_anonymous_class_exception
         end
       end
 
-      class ConnectionTestModel < ActiveRecord::Base
-        self.abstract_class = true
-      end
```
Comment (machty) [inline activerecord/test/cases/connection_pool_test.rb:496]: This test class has been split into two, one for thread tests and one for fiber, because there were some issues with running both test suites twice against the same class, some connection state that wasn't being totally cleaned up that I couldn't nail down. The failures weren't specific to the new Fiber code; simply running the threads test suite twice in a row produced the issue.

File: activerecord/test/cases/connection_pool_test.rb [hunk 6]
```diff
@@ -373,15 +380,15 @@ def test_checkout_fairness
         errors = []
 
         threads = expected.map do |i|
-          t = Thread.new {
+          t = new_thread {
             begin
               @pool.checkout # never checked back in
```
Comment (machty) [inline activerecord/test/cases/connection_pool_test.rb:385]: This line hangs when testing fibers, because without implementing a Ruby 3.1 scheduler that can hook into the blocking pool mutex wait, the fiber doesn't yield execution to any other fibers and simply hits the connection pool timeout.

**Comment by machty** [comment]
@byroot This is ready for another review pass.

I had a few doubts at first about my testing approach (literally the same shared tests with a few hooks to split b/w testing threads vs fibers), but I think it ended up being reasonably thorough with not too much of a diff or increase in complexity.

There are some specs that are explicitly skipped in fiber mode because they require a Ruby 3.1+ Fiber Scheduler to work (most cases involved a Fiber checking out a full pool and blocking until the timeout is hit). I figure it's probably too much to introduce in this PR but it seems at some point we'll want to add some faux Async task runner for Ruby 3.1+ and unskip these tests.

@ioquatix I remember some recent discussion about whether Ruby should offer some reference implementation for a scheduler; that might be relevant here?

File: activerecord/test/cases/connection_pool_test.rb [hunk 7]
```diff
@@ -27,8 +31,13 @@ def setup
...
 
-      teardown do
+      def skip_fiber_testing
+        skip "Can't test isolation_level=fiber without a Ruby 3.1+ Fiber Scheduler" if @isolation_level == :fiber
+      end
+
+      def shared_teardown
         @pool.disconnect!
+        ActiveSupport::IsolatedExecutionState.isolation_level = @previous_isolation_level
       end
```
Comment (byroot) [inline activerecord/test/cases/connection_pool_test.rb:41]: Nitpick, rather than `shared_teardown / shared_setup` it should just be `teardown / setup` and then call `super` (except for teardown since you have no more code to add).

File: activerecord/test/cases/connection_pool_test.rb [hunk 8]
```diff
@@ -772,5 +789,88 @@ def with_single_connection_pool
...
+      end
+
+      teardown do
+        shared_teardown
+      end
+
+      private
+        def new_thread(*args, &block)
+          Thread.new(*args, &block)
+        end
```
Comment (byroot) [inline activerecord/test/cases/connection_pool_test.rb:812]: Another nitpick
```suggestion
        def new_thread(...)
          Thread.new(...)
        end
```

File: activerecord/test/cases/connection_pool_test.rb [hunk 9]
```diff
@@ -772,5 +789,88 @@ def with_single_connection_pool
...
+          Thread.stop
+        end
+
+        def pass_to(_thread)
+          Thread.pass
+        end
+
+        def kill_current
+          Thread.current.kill
+        end
```
Comment (byroot) [inline activerecord/test/cases/connection_pool_test.rb:824]: It's a bit handwavy, but I'm scared that this may be called in the main thread causing minitest to exit 0 and to silently skip a large part of the test suite. Could we install call kill on the thread stored in a variable or received as argument instead? That would make me sleep better.

**Review by byroot — APPROVED** [review]
Some minor concerns but look good to me. Out of curiosity, how much work / code would a basic scheduler be to unskip these tests? Since we already test 3.1 we could likely include one in the test suite.

**Comment by machty** [comment]
@byroot PR feedback addressed in latest push.

> how much work / code would a basic scheduler be to unskip these tests?

Might be a better question for @ioquatix, I'm not sure how much work would be involved in getting a simple Fiber Scheduler-driven Async-lite working, but if it's not bad I wouldn't mind fixing up the currently-skipped tests.

**Review by byroot — APPROVED** [review]
LGTM, would be worth adding a CHANGELOG entry.

I'll wait a few days in case @matthewd has opinions.

**Comment by machty** [comment]
Changelog entry added.

File: activerecord/test/cases/connection_pool_test.rb [hunk 10]
```diff
@@ -772,5 +787,80 @@ def with_single_connection_pool
...
+
+    class ConnectionPoolFiberTest < ActiveRecord::TestCase
+      include ConnectionPoolTests
+
+      class FiberConnectionTestModel < ActiveRecord::Base
+        self.abstract_class = true
+      end
+
+      class ThreadlikeFiber < Fiber
+        class ThreadKilled < StandardError; end
```
Comment (matthewd) [inline activerecord/test/cases/connection_pool_test.rb:828]: Leftover?

File: activesupport/lib/active_support/isolated_execution_state.rb [hunk 11]
```diff
@@ -18,9 +18,9 @@ def isolation_level=(level)
         end
 
         if level != isolation_level
-          clear
-          singleton_class.alias_method(:current, "current_#{level}")
-          singleton_class.send(:private, :current)
+          clear if @isolation_level
+          singleton_class.alias_method(:executor, "current_#{level}")
```
Comment (matthewd) [inline activesupport/lib/active_support/isolated_execution_state.rb:22]: What do we think of making this more of a:

```ruby
self.scope = case level
             when :thread; Thread
             when :fiber; Fiber
             end
```

and then

```ruby
def context
  scope.current
end
```

(partly because I'd rather continue avoiding "executor"; partly because if we were going to use executor I feel like it more fits 'the thing that drives execution' than 'an execution', i.e. `Thread` not `Thread.current`; partly because I like the idea of removing this singleton class manipulation anyway)

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 12]
```diff
@@ -223,17 +223,17 @@ def valid_type?(type) # :nodoc:
       # this method must only be called while holding connection pool's mutex
       def lease
         if in_use?
-          msg = +"Cannot lease connection, "
-          if @owner == Thread.current
+          msg = + "Cannot lease connection, "
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:226]: ✂️  unary operator doesn't get a space

**Review by matthewd — APPROVED** [review]
This looks great!

Especially having lived in Thread-land for a long time now, I value keeping the tests operating in a faux-thread style, at least for now, and so naming is pretty immaterial there.

However, I do think the actual implementation needs some further renames. (Let's not keep a `@lock_thread` that can sometimes be a Fiber for too long.) Basically, I'd say it's worth purging all thread mentions from these `lib/` files. But admittedly that's hard while we're a bit up-in-the-air on what the best succinct name for `(Thread || Fiber)` is. Happy to bump that to a follow-up.

I know there are some detailed comments around locking considerations in here, too, and wonder what revisions they might need beyond `s/Thread/ThreadOrFiber/`. Are our actual locks all safe in a fiber world? (Obviously we worry less about other random code coming along, but do we have blocks whose assumption of functional atomicity could be broken by a fiber scheduler, because they do things that could switch fibers?)

File: activesupport/lib/active_support/isolated_execution_state.rb (hunk 11 above)
Comment (byroot) [reply activesupport/lib/active_support/isolated_execution_state.rb:22]: Ah yeah, 👍  to `scope`, that's a much better name than `executor`.

Comment (machty) [reply activesupport/lib/active_support/isolated_execution_state.rb]: I pushed another commit to resolve this, but just to be clear, there are few terms now in play based on this feedback (i.e. a few class methods, public or private, on `IsolatedExecutionState`):

- `scope` returns the Thread or Fiber class
- `context` returns the currently running instance of Thread or Fiber, which is now being used throughout my ConnectionPool changes as a reference to the current Thread or Fiber
- `state` returns the `active_support_execution_state` hash stashed on the `context`

**Comment by machty** [comment]
@matthewd I think punting some of the LockOrThread wordsmithing/refining to a followup makes sense. There's also [docs that reference "synchronizing thread access to a limited number of connections", etc](https://github.com/rails/rails/blob/9a1658cfcc86f418d2106a0d893a8de2bc80e3fa/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L47) that'll need to be addressed.

FYI the [wiki for Fibers](https://en.wikipedia.org/wiki/Fiber_(computer_science)) defines it as a "particularly lightweight thread of execution"; seems like the outside world also has trouble picking the right word.

File: activesupport/lib/active_support/isolated_execution_state.rb [hunk 13]
```diff
@@ -4,53 +4,58 @@
...
-          clear
-          singleton_class.alias_method(:current, "current_#{level}")
-          singleton_class.send(:private, :current)
-          @isolation_level = level
+        clear if @isolation_level
+
+        self.scope = case level
+                     when :thread; Thread
+                     when :fiber; Fiber
         end
```
Comment (matthewd) [inline activesupport/lib/active_support/isolated_execution_state.rb:28]: Ah, right. I believe Rubocop allows this variant a slightly tidier `end`:

```suggestion
        self.scope =
          case level
          when :thread; Thread
          when :fiber; Fiber
          end
```

File: activesupport/lib/active_support/isolated_execution_state.rb [hunk 14]
```diff
@@ -4,53 +4,58 @@
...
   module IsolatedExecutionState # :nodoc:
-    @isolation_level = :thread
+    @isolation_level = nil
 
     Thread.attr_accessor :active_support_execution_state
     Fiber.attr_accessor :active_support_execution_state
 
     class << self
       attr_reader :isolation_level
+      attr_accessor :scope
```
Comment (matthewd) [inline activesupport/lib/active_support/isolated_execution_state.rb:14]: Given `isolation_level` is the public API, probably best to make this a reader and set it directly as `@scope` below, actually.

File: activesupport/lib/active_support/isolated_execution_state.rb (hunk 11 above)
Comment (matthewd) [reply activesupport/lib/active_support/isolated_execution_state.rb]: Cool, thanks -- that matches my intention, but I'll let @byroot confirm he's happy with those names / API.

File: activesupport/lib/active_support/isolated_execution_state.rb (hunk 14 above)
Comment (machty) [reply activesupport/lib/active_support/isolated_execution_state.rb:14]: `scope` needs to be public so that the various locking mechanisms in the connection pool code can access the current thread/fiber; also, the reaper code checks `Fiber/Thread#alive?` directly.

Comment (machty) [reply activesupport/lib/active_support/isolated_execution_state.rb]: Gotcha, should be resolved now

**Comment by byroot** [comment]
Sorry for the late merge, it kinda slipped my mind this week.

Thanks for the PR!

**Comment by machty** [comment]
Thank you @byroot and @matthewd for the guidance! I didn't have a lot of confidence that I'd be able to get this through but you were both very helpful and encouraging.

**Comment by machty** [comment]
Here's a repo of an Edge Rails app that showcases this new feature https://github.com/machty/rails-edge-async-test

**Comment by machty** [comment]
@byroot I saw your [comment on the HN post](https://news.ycombinator.com/item?id=30456949) about this PR:

> > "Ideally, in the foreseeable future, we can expect good performance improvements to Rails I/O operations."

> Not really no. I/O will still be IOs, you may gain a tiny bit because of the cheaper "context switches", but it's gonna be very marginal for the vast majority of Rails apps.

> The goal is to allow the community to use fiber based server if they wish, but I (the author of most of these changes) don't except any substantial performance improvement from it.

Obviously Fibers aren't going to make IO operations complete more quickly, but isn't the argument for Fibers that:

1. If you have an IO-bound app, you want the cheapest concurrency primitive possible, both in terms of cheap context switches AND minimal memory usage?
2. Fibers are a lot less memory intensive than Threads (you can spin up way more Fibers than Threads before hitting memory limits)
3. The cooperative multitasking of Fibers can lead to much more deterministic tests, fewer tricky race conditions to debug

**Comment by byroot** [comment]
> 1. If you have an IO-bound app

That's the key here. I've never seen an IO bound Rails app. At best you may have something like 50% IO, in which case you slap 2 puma thread and call it a day. 2 threads vs 2 fibers is pretty much not measurable.

Now you could have some stripped down Rails app, doing mostly IO, and it make sense to accommodate them, but it's very marginal.

> Fibers are a lot less memory intensive than Threads

Same, this matter once you enter the 100, 1000 threads range. So that makes sense for a lightweight SSE / SebSocket or some kind of proxy mostly forwarding IOs, not for something that will render HTML templates etc.

> much more deterministic tests

I don't really see how this applies.

**Comment by trevorturk** [comment]
> That's the key here. I've never seen an IO bound Rails app. At best you may have something like 50% IO, in which case you slap 2 puma thread and call it a day. 2 threads vs 2 fibers is pretty much not measurable.
> 
> Now you could have some stripped down Rails app, doing mostly IO, and it make sense to accommodate them, but it's very marginal.

FWIW, I'd be happy to share more details, but I've migrated a Rails app that was using Puma/Typhoeus::Hydra to Falcon/Async::HTTP and now it needs to autoscale much less frequently. I'll write another blog post about the migration eventually, but you can get an idea of my use case here: https://helloweather.com/blog/inside-the-weather-machine -- tl;dr we proxy requests to upstream weather data APIs and reformat the responses into a unified format for the client. 

I created some demo apps when I was testing things here: https://github.com/socketry/async-examples which may be interesting. I'd also be happy to give a video call tour of my private app if that might be at all helpful or interesting to others. I worked with @ioquatix on the migration and I'm very happy with the results. I'm holding off on upgrading to Ruby 3 until everything is smoothed out a bit more, but I'm hopeful that I'll see additional savings once that's complete. 

One interesting thing I've seen is that my Heroku setup seems to be CPU ("dyno load") bound as opposed to memory bound now. I've considered adjusting my autoscaling setup to accommodate this (currently autoscaling based on queue time) but in Falcon's case, it seems queue time is almost always extremely low, unless the upstream APIs are performing particularly poorly, things don't back up much.

**Comment by byroot** [comment]
> tl;dr we proxy requests to upstream weather data APIs and reformat the responses into a unified format for the client.

Yep, exactly the kind of use case I was talking about. If you mostly proxy request and don't transform the response much, then you're really IO intensive. Your use case is what NodeJS and other evented frameworks were developed for, and it's great to be able to to that in Ruby in a much easier and elegant way than in the past, But that's the exception not the rule.

Mosts Rails apps don't look like this at all. 

I'm just trying to tone down people expectations, because when you read some comments on Reddit or HN, many seem to believe they'll just be able to replace `puma` by `falcon` in their gemfile and somehow get a perf boot. That's not the case.

**Comment by machty** [comment]
I was motivated to land this PR because my app is also IO bound -- a majority of our requests involve invoking an external service to evaluate ecommerce carts and recompute promotional discounts, which isn't a system we have any control over to speed up. It's fine that the requests are slow, but it's wasteful that we need to spin up more dynos and pay higher server bills to accommodate it.

I agree that it's good to temper folks' expectations, but it's also good to remove arbitrary barriers to folks' engineering creativity. It'd be great to just definitively say "Rails supports the same IO model that drives Node" rather than just conclude "Oh right, we can't use Rails for that".

**Comment by trevorturk** [comment]
Yep, all good points. I investigated using NodeJS and Elixir for my task, for example, but I was pleased to see that Ruby is increasingly capable here, so I was able to use the language I enjoy the most. I think "full stack" sort of Rails apps wouldn't benefit much from a switch, but I hope increased support for Fibers will let more people use Ruby for these sorts of task if they prefer! (Thank you, @machty!)

**Comment by byroot** [comment]
> but it's also good to remove arbitrary barriers

I introduced `IsolatedExecutionState` for this reason, so you're preaching to the choir. But:

>  to folks' engineering creativity.

I actually do fear folks engineering creativity, because keep in mind that IO heavy operations don't mix well with CPU heavy ones.  If you have a "traditional Rails app" that's relatively light on IO, switch to fibers and add some proxying endpoint or some websocket server in the same process, your latency will suffer terribly. That's what I fear too many people don't understand.

**Comment by trevorturk** [comment]
> IO heavy operations don't mix well with CPU heavy ones

I wonder if you could expand on this a bit, or perhaps suggest resources where people like me could learn more. For example, I have noticed that I seem to be CPU bound now (as opposed to memory-bound, which seems more common) and I didn't realize this might be a problem, and I'm not sure if I might be able to improve the situation without increasing my dynos back up! I expect people like me will run into this situation, then, and it's a bit out of the Rails mainstream knowledge, which I think is your point.

**Comment by byroot** [comment]
> I wonder if you could expand on this a bit

It all comes down to the GVL. You can also search for articles on Python's GIL, it's exactly the same concern. There's also probably plenty of articles on why you should avoid CPU intensive operations in Node, but I'm much less familair with it.

In short only one Fiber can execute at a time, just like threads, but contrary to threads they're not preempted until they do some IO.

So if you have a controller action that execute Ruby for say 25ms without ever doing IOs, e.g.:

```ruby
def fibonnaci
  render plain: fibonacci(100)
end
```

No other fiber will be able to execute during that time.

So if in the same process you have some lightweight WebSocket server that is just forwarding messages it gets from a Redis subscription, whenever someone hits the 25ms fibbonacci endpoint, your WebSocket server won't be able to forward any message for 25ms.

The fibonnaci thing is of course a silly example, but if you look at a typical Rails app, you'll easily see request spending more time than that rendering ERB templates or serializing JSON.

That's why you want to isolate both use cases in their respective process, use a threaded server (or even just no threads at all) for the bread and butter transactional Rails workload, and a fiber server for your WebSocket stuffs or light proxying stuff.

**Comment by trevorturk** [comment]
Thank you, this is very informative and helpful!

**Comment by simi** [comment]
> IsolatedExecutionState seemed to be a step in the direction of abstracting whether the "current running thing" is a Thread or a Fiber, but as I look around the Rails codebase, I see lots of code that is hard-wired to a Thread-centric implementation. The more we push in the direction of making Rails Async/Fiber-friendly, the more we'll run into framework code that's hardwired to Thread-centric APIs, some examples:
> 
> * [ActionCable::Connection::StreamEventLoop uses a Thread and ThreadPoolExecutor to implement an event loop](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/actioncable/lib/action_cable/connection/stream_event_loop.rb#L69-L75)
> * [The ActionCable PostgreSQL subscription adapter listens to async events via a thread running in a loop](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/actioncable/lib/action_cable/subscription_adapter/postgresql.rb#L82-L85)
> * [The ActionCable Redis subscription handler does a similar thing](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/actioncable/lib/action_cable/subscription_adapter/redis.rb#L150-L155)
> * [ActionController::Live spins up a thread to stream the response](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/actionpack/lib/action_controller/metal/live.rb#L339-L345)
> * [ActiveJob's in-process AsyncAdapter uses a ThreadPoolExecutor](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/activejob/lib/active_job/queue_adapters/async_adapter.rb#L89)
>   
>   * this maybe fine; we could probably just create another Fiber-friendly adapter?
> * [load_async is driven either by a global or per connect ThreadPoolExecutor](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/activerecord/lib/active_record.rb#L219)
> * [ActiveRecord::ConnectionAdapters::TransactionState check Thread#status == 'aborting' and there is no Fiber#status](https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L336)
> 
> ...and more.
> 

Another place I have found out facing similar problem is ActiveRecord lock used around transactions.

https://github.com/rails/rails/blob/9062d705ad9055c0f3ec73505150e0e94fff5925/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L104

It is using `LoadInterlockAwareMonitor` which is slightly modified `Monitor` from `stdlib`. Ruby 3.0.2 was shipped with bugfix (https://github.com/ruby/ruby/commit/f9196de1dee6f5ab8b6fe115070b92775a3500fe, https://bugs.ruby-lang.org/issues/17827) to make `Monitor` fiber-safe. That means starting Ruby 3.0.2 ActiveRecord code related to this lock is also fiber safe, even (if I understand it well) thread safety would be enough here. It is causing problem with wrapping ActiveRecord calls in `Enumerator`.

It seems similar to original reason to introduce `IsolatedExecutionState` at https://github.com/rails/rails/commit/540d2f41f6913cd6c5a71301540bfe1551c2acc5.

Following code hangs on 3.0.2 (works well on older Rubies).

```ruby
ApplicationRecord.transaction { Enumerator.new { |y| y.yield ApplicationRecord.connection.execute("SELECT 1") }.next }
```

Since I wasn't able to find out any replacement for `Monitor` being `Thread` oriented and since those transaction locks in AR were introduced to handle race condition in parallel tests which I'm not using (since I'm running rspec in our app) I'm currently testing bypassing this lock with following monkey-patch.

```ruby
class SimplifiedLoadInterlockMonitor < ActiveSupport::Concurrency::LoadInterlockAwareMonitor
  def mon_try_enter
    true
  end

  def mon_exit
    true
  end
end

class ActiveRecord::ConnectionAdapters::AbstractAdapter
  attr_writer :lock
end

ActiveRecord::ConnectionAdapters::AbstractAdapter.set_callback :checkout, :before, ->(connection) {
  connection.lock = SimplifiedLoadInterlockMonitor.new
}
```

Is this something to be fixed in `ActiveRecord` itself?

If welcomed, I can also contribute ability to specify own lock implementation for `ActiveRecord` (just the `attr_writer`).

**Comment by trevorturk** [comment]
@simi this is cool to see, I think the main problem we have with Fibers and Rails at the moment is that nobody has taken the time to list the problem areas, so we have some unknowns. 

For example, I'm using Falcon and Async::HTTP with the latest released version of Rails and haven't seen any issues. I'm not using ActiveRecord yet, or ActionCable etc, so perhaps I'm just getting lucky so far. 

I wonder if you want to try to work up a complete list of where you think we'd need updates, by subsystem. So, if I'm not working with ActionCable, I could ignore those and focus on the ActiveRecord bits etc. 

If we had a work list, I bet we could encourage some more contributions to improve support.

**Comment by simi** [comment]
@trevorturk I think it is not easily possible to scan a code base to find all possible problems. But it should be possible to use my findings as a base. Anything using locking based on `Monitor` could potentially face the same problem if calls are nested and custom iteration on `Enumerator` is used.

Any idea what to do to properly handle my mentioned example usecase? Should we introduce locking mechanism based on `IsolationExecutionState` configuration?

**Comment by ioquatix** [comment]
> Any idea what to do to properly handle my mentioned example usecase? Should we introduce locking mechanism based on IsolatedExecutionState configuration?

I'd personally vote no, it's a bad idea to change the semantics of a program in this way.

**Comment by byroot** [comment]
> `Enumerator`

This one unfortunately I really don't know how to solve, and I think it's a common problem ruby-core is aware of too and doesn't know how to fix either.

A side goal for `IsolatedExecutionState` was to replace a lot of fiber locals (`Thread.current[:var]`) by actual thread locals (`Thread.current.thread_variable_get`), as to fix the use of `Enumerator` for people running threaded servers.

However, I have no idea how `Enumerator` could work properly on fiber servers, short of having some kind of API to copy or share the Fiber locals with the newly spawned fiber.

**Comment by eregon** [comment]
FWIW, Enumerator is only problematic when `Enumerator#{next,next_values,peek,peek_values}` are used.
In addition, Enumerable#zip and Enumerator#zip use `Enumerator#next` when given non-Array arguments.
All other Enumerator/Enumerable methods are fine and do not create a Fiber.

> other internal Ruby fibers?

There are no other internal Ruby fibers except than for the above Enumerator methods.

**Comment by ioquatix** [comment]
> However, I have no idea how Enumerator could work properly on fiber servers, short of having some kind of API to copy or share the Fiber locals with the newly spawned fiber.

That's now possible using inheritable fiber storage, not sure if it completely solves the problem but maybe worth a shot.
